Rewrite

po5 opened the pull request:
Should make it easier to add new features. Much less dumb global state management required.

Finally closes #2. Closes #5. I've realized there probably is no downside to restoring original speed, didn't make an option to disable it.
Closes #1 through a script message to be used for input.conf commands like `; script-message-to evafast set-option speed_cap 5`.
Closes #7.

All current functionality and dumb QoL stuff I could remember has been reimplemented.  
~~It's not impossible I've missed edge cases.~~ I have missed edge cases.

`speedup-target` is barely tested, but seems to work fine.  
Subtitle lookahead should be slightly more accurate.

To do:
- [x] Remove hardcoded `1`'s in `speed_transition()`, maybe use math instead of dumb recursion
- [x] Better caching/clearing of `next_sub_at`
- [x] Better displayed subtitle heuristic, check `sub-start` for image-based subs and `sub-text` for text-based subs
- [x] Flatten if pyramids in `flash_speed()`
- [ ] Redo #4
- [ ] Steal `dropOnAVdesync` from speed-transition.lua
- [ ] Warn about low `demuxer-readahead-secs` and `cache-secs` (depending on current settings and/or automatically set them)
- [ ] Document script-bindings better
- [ ] Customizable seek command #6
- [ ] Reset speed when seeking past speedup-target
- [x] Rewinding while toggled
- [x] Rewind toggle binding
- [ ] Subtitle lookbehind when rewinding
- [ ] Solve speed fluctuation at gap between subtitles in slowdown range (including when the player is paused)
- [ ] Restore original play-dir, if the user was playing a video backwards for whatever reason
- [x] Releasing speedup key after reaching full speed on files without subtitles shouldn't seek
- [ ] (?) Simulate repeat to allow for mouse input #3
- [ ] Ability to use osd display even when the user has uosc
- [ ] Holding down a toggle key shouldn't make speed flicker
- [x] Don't fetch next sub for every single line in frame-by-frame typesetting

po5 committed:
Rewrite beginnings
Should make it easier to add new features

po5 commented on evafast.lua:
```diff
@@ -43,244 +43,243 @@ local options = {
+                if next_sub_at < current_time then
+                    next_sub_at = next_sub(current_time)
+                end
+                if next_sub_at >= current_time then
```
`>` instead of `>=`?

po5 commented on evafast.lua:
```diff
@@ -43,244 +43,243 @@ local options = {
+                end
+                if next_sub_at >= current_time then
+                    local time_for_correction = speed_transition(current_speed, options.subs_speed_cap)
+                    if (time_for_correction * current_speed + current_time) > next_sub_at then
```
`>=` instead of `>`?

po5 commented on evafast.lua:
```diff
@@ -43,244 +43,243 @@ local options = {
-            speedup_target = nil
-        end
-        if speed_timer ~= nil and not toggle_state and mp.get_property_number("speed") == 1 and ((not options.subs_speed_cap or mp.get_property("sub-start") == nil) and options.speed_cap or options.subs_speed_cap) ~= 1 then
+    if math.floor(target_speed * 1000) == math.floor(current_speed * 1000) then
```
I hate this, but simple comparison of floats that should be the same doesn't work here.  
Look for a better solution, or benchmark this vs `math.abs(target_speed - current_speed) < 0.00001` vs `string.format()`.

po5 commented:
Does this work as expected for you @tomasklaen, @xfzv?

po5 committed:
Rewind feature

po5 commented:
Lazy implementation of #7. Would have to make the subtitle lookahead work in reverse as well, seems doable.

```patch
diff --git a/evafast.lua b/evafast.lua
index 72efda3..b78bd22 100644
--- a/evafast.lua
+++ b/evafast.lua
@@ -57,6 +57,7 @@ local toggled = false
 local speedup = false
 local original_speed = 1
 local next_sub_at = -1
+local rewinding = false
 
 local function speed_transition(test_speed, target_speed)
     local speed_correction = 0
@@ -207,6 +208,10 @@ local function adjust_speed()
             speed_timer:kill()
             toggled_display = true
             speedup_target = nil
+            if rewinding then
+                mp.set_property("play-dir", "+")
+                rewinding = false
+            end
         end
         return
     end
@@ -234,7 +239,11 @@ end
 speed_timer = mp.add_periodic_timer(100, adjust_speed)
 speed_timer:kill()
 
-local function evafast(keypress)
+local function evafast(keypress, rewind)
+    if rewinding and not rewind then
+        rewinding = false
+        mp.set_property("play-dir", "+")
+    end
     if keypress["event"] == "down" then
         if not speed_timer:is_enabled() then
             if not toggled then
@@ -258,7 +267,13 @@ local function evafast(keypress)
         else
             evafast_speedup()
         end
-        mp.commandv("seek", options.seek_distance)
+        if rewind then
+            mp.commandv("seek", -options.seek_distance)
+            mp.set_property("play-dir", "+")
+            rewinding = false
+        else
+            mp.commandv("seek", options.seek_distance)
+        end
         flash_speed()
     elseif keypress["event"] == "repeat" or keypress["event"] == "up" then
         speedup = toggled or keypress["event"] == "repeat"
@@ -266,12 +281,20 @@ local function evafast(keypress)
             toggled_display = false
         end
         if speed_timer:is_enabled() then return end
+        if rewind then
+            mp.set_property("play-dir", "-")
+            rewinding = true
+        end
         speed_timer.timeout = 0
         speed_timer:resume()
         speed_timer.timeout = options.speed_interval
     end
 end
 
+local function evafast_rewind(keypress)
+    evafast(keypress, true)
+end
+
 mp.observe_property("sid", "native", function(prop, val)
     has_subtitle = (val or 0) ~= 0
     next_sub_at = -1
@@ -315,6 +338,7 @@ end)
 
 
 mp.add_key_binding("RIGHT", "evafast", evafast, {repeatable = true, complex = true})
+mp.add_key_binding(nil, "evafast-rewind", evafast_rewind, {repeatable = true, complex = true})
 mp.add_key_binding(nil, "flash-speed", function() flash_speed(mp.get_property_number("speed", 1), nil, true) end)
 mp.add_key_binding(nil, "speedup", evafast_speedup)
 mp.add_key_binding(nil, "slowdown", evafast_slowdown)

```

xfzv commented:
> Does this work as expected for you @tomasklaen, @xfzv?

Yes, all good here. Solves #2.

tomasklaen commented:
I wrote my own little speed adjustment script thingy, so don't wait on me with this :)

po5 commented:
Please share that, always good to spread ideas (:

tomasklaen commented:
It's just a dumb hold to speed up or toggle a speed up. Don't want to maintain it so here's just a pastebin: https://pastebin.com/acXpcVYR

xfzv commented:
As far as I'm concerned, due to #2 , I've been using this along with [Natural-Harmonia-Gropius/InputEvent](https://github.com/Natural-Harmonia-Gropius/InputEvent):

```
RIGHT               seek 5 ; script-binding uosc/flash-timeline             #@click
RIGHT               no-osd set speed 3.5 ; script-binding uosc/flash-speed  #@press
RIGHT               ignore                                                  #@release
```

until the rewrite is merged into the main branch.

po5 committed:
Simplify speed_transition, remove 1x limitation

po5 committed:
Flatten flash_speed

po5 committed:
Reorder conditions

po5 committed:
Speedup-target improvements

po5 committed:
Update time_for_correction formula

po5 committed:
Subtitle lookahead when subs are displayed

po5 committed:
Relative speedup-target
+ sign is relative to current time, - sign is relative to video end
Former is useful for "speedup for n seconds" keybind
Latter is useful for catching up on livestreams

po5 committed:
Rework key handling and timer management, subs correction improvements
Need to re-add rewinding

po5 committed:
Ensure timer on external calls to slowdown

po5 committed:
Clear next sub on seek

po5 committed:
Reintroduce rewind

po5 committed:
Account for speedup-target in more situations

po5 committed:
Rewind toggle

po5 committed:
Take adjusted speed into account for transition time

po5 committed:
Don't trigger sub speed cap on empty lines for text subtitles

po5 committed:
Reset play-dir when cancelling rewind toggle with regular toggle keybind

po5 committed:
Account for multiply_modifier in speed_transition

po5 commented:
Current state of affairs.

Slowdown with `lookahead = true` was wrong since its inception and would kick in slightly too late. That has been fixed.  
`sub-visibility` flicker should also be reduced.  
Haven't ran into fluctuating speed issues since making the lookahead accuracy fix, but I doubt it's fixed the underlying issue.

`show_*` options are a bit silly, maybe a single `show_speed` option with a comma-delimited list of which combinations of displays+states we allow e.g. `osd_speed,uosc_speed,osc_toggled,uosc_toggled,osd_target`. I'm afraid this would confuse people.

The functionality that's there has been tested enough that I'm contemplating making this the default branch, until it's finished and gets merged. Will do after documentation is written.

po5 committed:
Don't run lookahead on every line of frame-by-frame typesetting
This is an accuracy tradeoff - not knowing the next sub right away
can cause speed flicker but I haven't noticed real-world regressions

christoph-heinrich commented:
I have used this for about a week now and encountered two times where it would constantly set the speed to 2 (the original speed). Holding the forward button accelerates as it should and after letting go it decelerates again, but after reaching 2 it doesn't stop setting the speed to 2. As a result I can't change the speed because it will be immediately set to 2 again.

Unfortunately I couldn't figure out how to trigger it, but it's annoying when it happens.

<details>
<summary>Here is my conf in case it helps.</summary>

```
# How far to jump on press
seek_distance=5

# Playback speed modifier, applied once every speed_interval until cap is reached
speed_increase=0.07
speed_decrease=0.035

# At what interval to apply speed modifiers
speed_interval=0.05

# Playback speed cap
speed_cap=3

# Playback speed cap when subtitles are displayed, 'no' for same as speed_cap
subs_speed_cap=1.6

# Multiply current speed by modifier before adjustment (exponential speedup)
# Use much lower values than default e.g. speed_increase=0.05, speed_decrease=0.025
multiply_modifier=yes

# Show current speed on the osd (or flash speed if using uosc)
show_speed=yes

# Show current speed on the osd when toggled (or flash speed if using uosc)
show_speed_toggled=yes

# Show seek actions on the osd (or flash timeline if using uosc)
show_seek=yes

# Look ahead for smoother transition when subs_speed_cap is set
lookahead=yes
```
</details>

po5 commented:
I've encountered this multiple times as well, and it killed my momentum. Haven't looked too much into it.

xfzv commented:
@christoph-heinrich 

What you describe seems similar to https://github.com/po5/evafast/issues/9

Doesn't it?

christoph-heinrich commented:
@xfzv Yes that's it, I didn't check issues because I didn't expect there to be one for a PR.

christoph-heinrich committed:
Round values in timer target speed check

math.floor() can result in one value being off by 1 because one
was just barely below the whole integer while the other one was not.
(floating point numbers \o/)
Adding 0.5 for actual rounding avoids that problem.

Closes #9

po5 committed:
Add note about regex filter for ignoring empty subtitle lines
This gets rid of text subtitle detection and our own empty subtitle
check, since it didn't work with subs_lookahead for muxed subtitles
(sub-text wasn't available straight away).

po5 committed:
More robust regex filter
Previous regex would skip subtitles that contained empty lines but
weren't empty. This is POSIX Extended RegEx, doesn't seem possible
in plain POSIX or mpv's JS regex to match the start and end of a
multiline string.

po5 committed:
Only fetch current speed value in flash-speed binding when necessary

verygoodlee commented:
`evafast.conf` has a key `lookahead`, but in`evafast.lua` named `subs_lookahead` 
![image](https://github.com/po5/evafast/assets/45035465/78dc8a1c-e4a0-4530-871f-8b29086fb80d)

po5 committed:
Update sample config

po5 committed:
Update options through script-opts, remove set-option message

po5 commented:
2c00ab232d0c7db43ebc4634f386756bde3dd041 removed our empty subtitle check, because `sub-text` wasn't available in time.  
I wonder if anything improved after https://github.com/mpv-player/mpv/pull/12819. It apparently helped with @christoph-heinrich's [mpv-subtitle-lines](https://github.com/christoph-heinrich/mpv-subtitle-lines).

christoph-heinrich commented:
I'd expect it to help here too.

verygoodlee commented:
There is a better symbol to replace `▶▶`, mpv builtin fastforward osd symbol.
![image](https://github.com/po5/evafast/assets/45035465/731fb69c-7c61-4161-be8b-0467ea7c85e7)
The definition in mpv source code is `OSD_FFW`, and it starts from `OSD_CODEPOINTS`.
See https://github.com/mpv-player/mpv/blob/6e161ff13fb83be9c5180e99427651975d33debf/sub/osd.h#L109-L133
the codepoint is `OSD_CODEPOINTS + OSD_FFW = 0xE000 + 0x05 = 0xE005`, \
but it's not a normal unicode codepoint, only the `mpv-osd-symbols` font can display it properly. 

`evafast.lua` can be written like this
```lua
local symbol = '▶▶'
if mp.get_property('osd-font') == 'mpv-osd-symbols' then symbol = '\u{E005}' end
mp.osd_message(symbol .. string.format(" x%.1f", current_speed))
```
`mpv.conf` change the  OSD font to `mpv-osd-symbols`
```
osd-font=mpv-osd-symbols
```
- Before
![image](https://github.com/po5/evafast/assets/45035465/e1334616-0334-4acf-9204-675663e3d435)
- After
![image](https://github.com/po5/evafast/assets/45035465/3fa67289-4785-497b-92e4-8cc5d23d8116)

po5 commented:
Thanks for bringing this up. We can use the symbol regardless of the user's osd-font with the ass \fn tag.

verygoodlee commented:
> We can use the symbol regardless of the user's osd-font with the ass \fn tag.

`mp.osd_message()` supports ass tag using [osd-ass-cc/{0|1}](https://mpv.io/manual/stable/#command-interface-osd-ass-cc)
```lua
local ass_start = mp.get_property_osd("osd-ass-cc/0")
local ass_stop = mp.get_property_osd("osd-ass-cc/1")
local symbol = "{\\fnmpv-osd-symbols}\u{E005}"
mp.osd_message(ass_start .. symbol .. ass_stop .. string.format(" x%.1f", current_speed))
```

po5 committed:
Customizable osd symbol, default to mpv's fast-forward icon
Still respects osd font for the main text
As suggested by @verygoodlee
https://github.com/po5/evafast/pull/8#issuecomment-1818184400

po5 committed:
Add rewind symbol
